src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,345 @@
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Internal;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    public class MultipartPipeReader
+    {
+
+        public const int DefaultHeadersCountLimit = 16;
+        public const int DefaultHeadersLengthLimit = 1024 * 16;
+        //private const int DefaultBufferSize = 1024 * 4;
+        private const int StackAllocThreshold = 128;
+
+        private readonly PipeReader _pipeReader;
+        //private ReadOnlySequence<byte> _buffer;
+        private long _bytesConsumed = 0;
+        private readonly MultipartBoundary _boundary;
+        private MultipartPipeReaderStream _currentStream;
+
+        private static readonly byte[] ColonDelimiter = new byte[] { (byte)':' };
+        private static readonly byte[] CrlfDelimiter = new byte[] { (byte)'\r', (byte)'\n' };
+
+
+        public MultipartPipeReader(string boundary, PipeReader pipeReader)
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+
+            _currentStream = new MultipartPipeReaderStream(_pipeReader, _boundary);
```

Comment:
Include comments and limit from:
https://github.com/aspnet/AspNetCore/blob/1189a2c294df55d7a67f1bc6ea74d74858f467bd/src/Http/WebUtilities/src/MultipartReader.cs#L48-L50